Deprecate the public definition of ERR_STATE

levitte opened the pull request:
The intention is to make it opaque later on.

levitte committed:
Deprecate the public definition of ERR_STATE

The intention is to make it opaque later on.

levitte commented:
This is a step toward #9082, which is marked to happen in 4.0

kroeckx commented on include/openssl/err.h:
```diff
@@ -38,19 +38,9 @@ extern "C" {
-    int top, bottom;
-} ERR_STATE;
+# if !OPENSSL_API_3
+#  include <openssl/err_state.h>
```
I don't understand why you make a new public file to deprecated something. Why don't you just put the `#if` around it, instead of moving it to a new file?

levitte replied:
Because `crypto/err/err.c` still needs the definition.  The easiest was to place it in another file that is unguarded.  Th other solution would be to duplicate the definition in a private header, but that would be a bad solution.

levitte replied:
Another solution is to have an internal macro defined by `crypto/err/err.c`, say `OPENSSL_FORCEDEF_ERR_STATE`, which overrides deprecation.  Is that preferable?

richsalz replied:
I was typing that in my browser when your reply showed up.  :) FWIW, *I* think that's preferable.

levitte replied:
Okie.  Maybe...  Yesterday, I thought *that* was a bad idea, but I'm less sure today.

levitte replied:
Done

kroeckx commented:
You also need to deprecate ERR_get_state

levitte commented:
> You also need to deprecate ERR_get_state

Ah, right

levitte committed:
Deprecate ERR_get_state()

Internally, we still need this function, so we make it internal and
then add a new ERR_get_state() that simply calls the internal variant,
unless it's "removed" by configuration.

levitte commented:
> You also need to deprecate ERR_get_state

Done

levitte commented:
I had forgotten `make update`

mattcaswell commented on include/openssl/err.h:
```diff
@@ -301,7 +303,7 @@ int ERR_load_ERR_strings(void);
 DEPRECATEDIN_1_1_0(void ERR_remove_thread_state(void *))
 DEPRECATEDIN_1_0_0(void ERR_remove_state(unsigned long pid))
-ERR_STATE *ERR_get_state(void);
+DEPRECATEDIN_3(ERR_STATE *ERR_get_state(void))
```
Seems like this deserves a CHANGES entry

levitte replied:
Yes, thanks for the reminder

levitte committed:
Document the deprecation of ERR_STATE and ERR_get_state()

mattcaswell reviewed (approved):
Reconfirm

levitte commented:
Merged.

14e275e8fb736be4ea83441b630515f7be97d06b Deprecate the public definition of ERR_STATE
e5d4233fbd07eac52227c7ec5f479a46f15914bf Deprecate ERR_get_state()
3c90534830fdb0cc0e790dd1d5fa2a90e1375e87 Document the deprecation of ERR_STATE and ERR_get_state()
